libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,834 @@
+    }
+    // Exclusively send Burst Data,
+    // We take over the transmit part of the link when sending FTP message,
+    // To avoid buffer corruption and maximise throughput
+    GCS_MAVLINK::lock_channel(_ftp_chan, false, true);
+    do {
+        more_data = false;
+        FTPErrorCode error_code = kErrNone;
+
+        stream_payload->seq_number = _session_info.stream_seq_number;
+        stream_payload->session = 0;
+        stream_payload->opcode = kRspAck;
+        stream_payload->req_opcode = kCmdBurstReadFile;
+        stream_payload->offset = _session_info.stream_offset;
+        //reset the data
+        memset(stream_payload->data, 0, MAX_FTP_DATA_LENGTH);
+        _session_info.stream_seq_number++;
+
+
+        // We have to test seek past EOF ourselves, lseek will allow seek past EOF
+        if (_session_info.stream_offset >= _session_info.file_size) {
+            error_code = kErrEOF;
+        }
+
+        if (error_code == kErrNone) {
+            if (lseek(_session_info.fd, stream_payload->offset, SEEK_SET) < 0) {
+                _ftp_errno = errno;
+                error_code = kErrFailErrno;
+            }
+        }
+
+        if (error_code == kErrNone) {
+            int bytes_read = ::read(_session_info.fd, &stream_payload->data[0], MAX_FTP_DATA_LENGTH);
+
+            if (bytes_read < 0) {
+                // Negative return indicates error other than eof
+                _ftp_errno = errno;
+                error_code = kErrFailErrno;
+            } else {
+                stream_payload->size = bytes_read;
+                _session_info.stream_offset += bytes_read;
+                _session_info.stream_chunk_transmitted += bytes_read;
+            }
+        }
+
+        if (error_code != kErrNone) {
+            stream_payload->opcode = kRspNak;
+            stream_payload->size = 1;
+            stream_payload->data[0] = error_code;
+
+            if (error_code == kErrFailErrno) {
+                stream_payload->size = 2;
+                stream_payload->data[1] = _ftp_errno;
+            }
+
+            _session_info.stream_download = false;
+
+        } else {
+            more_data = true;
+            stream_payload->burst_complete = false;
+        }
+
+        if (_session_info.stream_chunk_transmitted > _ftp_stream_chunk_size) {
+            stream_payload->burst_complete = true;
+            _session_info.stream_download = false;
+            _session_info.stream_chunk_transmitted = 0;
+            more_data = false;
+        }
+        _ftp_response.target_system = _session_info.stream_target_system_id;
+        _ftp_wait_until_buffer_free(stream_payload);
+        mavlink_msg_file_transfer_protocol_send_struct(_ftp_chan, &_ftp_response);
+        //also send heartbeat every 10 pkts arbitrarily to keep GCS's happy
+        if (!(_session_info.stream_chunk_transmitted % 10)) {
+            gcs().chan(_ftp_chan).send_heartbeat();
+        }
+    } while (more_data && _session_info.stream_download);
```

Comment:
Is this truly blocking all outgoing MAVLink until `_ftp_stream_chunk_size` has been sent out? That's a massive interruption to MAVLink service...